vcd/resource_vcd_vm_internal_disk.go

```diff
@@ -0,0 +1,475 @@
+	vcdClient := meta.(*VCDClient)
+
+	vcdClient.lockParentVm(d)
+	defer vcdClient.unLockParentVm(d)
+
+	vm, vdc, err := getVm(vcdClient, d)
+	if err != nil {
+		return err
+	}
+
+	var storageProfilePrt *types.Reference
+	var overrideVmDefault bool
+
+	if storageProfileName, ok := d.GetOk("storage_profile"); ok {
+		storageProfile, err := vdc.FindStorageProfileReference(storageProfileName.(string))
+		if err != nil {
+			return fmt.Errorf("[vm creation] error retrieving storage profile %s : %s", storageProfileName, err)
+		}
+		storageProfilePrt = &storageProfile
+		overrideVmDefault = true
+	} else {
+		storageProfilePrt = vm.VM.StorageProfile
+		overrideVmDefault = false
+	}
+
+	iops := int64(d.Get("iops").(int))
+	isThinProvisioned := d.Get("thin_provisioned").(bool)
+
+	diskSetting := &types.DiskSettings{
+		SizeMb:              int64(d.Get("size_in_mb").(int)),
+		UnitNumber:          d.Get("unit_number").(int),
+		BusNumber:           d.Get("bus_number").(int),
+		AdapterType:         internalDiskBusTypes[d.Get("bus_type").(string)],
+		ThinProvisioned:     &isThinProvisioned,
+		StorageProfile:      storageProfilePrt,
+		Iops:                &iops,
+		VirtualQuantityUnit: "byte",
+		OverrideVmDefault:   overrideVmDefault,
+	}
+
+	err = powerOffIfNeeded(d, vm)
+	if err != nil {
+		return err
+	}
+
+	diskId, err := vm.AddInternalDisk(diskSetting)
+	if err != nil {
+		return fmt.Errorf("error updating VM disks: %s", err)
+	}
+
+	d.SetId(diskId)
+
+	err = powerOnIfNeeded(d, vm)
+	if err != nil {
+		return err
+	}
+
+	return resourceVmInternalDiskRead(d, meta)
+}
+
+func powerOnIfNeeded(d *schema.ResourceData, vm *govcd.VM) error {
+	vmStatus, err := vm.GetStatus()
+	if err != nil {
+		return fmt.Errorf("error getting VM status before ensuring it is powered on: %s", err)
+	}
+
+	if (vmStatus != "POWERED_ON" && d.Get("bus_type").(string) == "ide" && d.Get("allow_vm_reboot").(bool)) ||
+		(vmStatus != "POWERED_ON" && d.Get("allow_vm_reboot").(bool) && (d.HasChange("bus_number") || d.HasChange("unit_number"))) {
+		log.Printf("[DEBUG] Powering on VM %s after adding internal disk.", vm.VM.Name)
+
+		task, err := vm.PowerOn()
+		if err != nil {
+			return fmt.Errorf("error powering on VM for adding/updating internal disk: %s", err)
+		}
+		err = task.WaitTaskCompletion()
+		if err != nil {
+			return fmt.Errorf(errorCompletingTask, err)
+		}
+	}
+	return nil
+}
+
+func powerOffIfNeeded(d *schema.ResourceData, vm *govcd.VM) error {
+	vmStatus, err := vm.GetStatus()
+	if err != nil {
+		return fmt.Errorf("error getting VM status before ensuring it is powered on: %s", err)
+	}
+
+	if (vmStatus != "POWERED_OFF" && d.Get("bus_type").(string) == "ide" && d.Get("allow_vm_reboot").(bool)) ||
+		(vmStatus != "POWERED_OFF" && d.Get("allow_vm_reboot").(bool) && (d.HasChange("bus_number") || d.HasChange("unit_number"))) {
+		log.Printf("[DEBUG] Powering off VM %s for adding/updating internal disk.", vm.VM.Name)
```

Comment:
I know this is not for review yet, but writing early. Please add test coverage for the above logic. Tests which would check successful disk management with:
1. IDE disk and allow_vm_reboot
2. Non-IDE disk without changing bus_number and unit_number and without allow_vm_reboot
3. Non-IDE disk with changing bus_number or unit_number and with allow_vm_reboot

Reply:
I can do that - I estimate additional ~10min to runing our test time if you think it's ok

Reply:
I guess we need it. We could exclude it with some tag like `skipLong` or so, but it is hard to be sure it works as expected. Will be even harder in future when we have to modify it

Reply:
Test for update and IDE are added.